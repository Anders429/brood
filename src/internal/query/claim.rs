use crate::{component::Component, entity, query::view};
use core::any::TypeId;
use hashbrown::HashSet;

pub trait Claim {
    fn claim(mutable_claims: &mut HashSet<TypeId>, immutable_claims: &mut HashSet<TypeId>);
}

impl<C> Claim for &C
where
    C: Component,
{
    fn claim(_mutable_claims: &mut HashSet<TypeId>, immutable_claims: &mut HashSet<TypeId>) {
        immutable_claims.insert(TypeId::of::<C>());
    }
}

impl<C> Claim for &mut C
where
    C: Component,
{
    fn claim(mutable_claims: &mut HashSet<TypeId>, _immutable_claims: &mut HashSet<TypeId>) {
        mutable_claims.insert(TypeId::of::<C>());
    }
}

impl<C> Claim for Option<&C>
where
    C: Component,
{
    fn claim(_mutable_claims: &mut HashSet<TypeId>, immutable_claims: &mut HashSet<TypeId>) {
        immutable_claims.insert(TypeId::of::<C>());
    }
}

impl<C> Claim for Option<&mut C>
where
    C: Component,
{
    fn claim(mutable_claims: &mut HashSet<TypeId>, _immutable_claims: &mut HashSet<TypeId>) {
        mutable_claims.insert(TypeId::of::<C>());
    }
}

impl Claim for entity::Identifier {
    fn claim(_mutable_claims: &mut HashSet<TypeId>, _immutable_claims: &mut HashSet<TypeId>) {}
}

impl Claim for view::Null {
    fn claim(_mutable_claims: &mut HashSet<TypeId>, _immutable_claims: &mut HashSet<TypeId>) {}
}

impl<'a, V, W> Claim for (V, W)
where
    V: Claim,
    W: Claim,
{
    fn claim(mutable_claims: &mut HashSet<TypeId>, immutable_claims: &mut HashSet<TypeId>) {
        V::claim(mutable_claims, immutable_claims);
        W::claim(mutable_claims, immutable_claims);
    }
}