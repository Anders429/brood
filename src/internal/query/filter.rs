use crate::{
    component::Component,
    entity,
    query::{filter::{And, Filter, Has, None, Not, Or}, view, view::{View, Views}},
};
use core::any::TypeId;
use hashbrown::HashMap;

pub trait FilterSeal {
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool;
}

impl FilterSeal for None {
    unsafe fn filter(_key: &[u8], _component_map: &HashMap<TypeId, usize>) -> bool {
        true
    }
}

impl<C> FilterSeal for Has<C>
where
    C: Component,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        match component_map.get(&TypeId::of::<C>()) {
            Some(index) => key.get_unchecked(index / 8) & (1 << (index % 8)) != 0,
            Option::None => false,
        }
    }
}

impl<F> FilterSeal for Not<F>
where
    F: Filter,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        !F::filter(key, component_map)
    }
}

impl<F1, F2> FilterSeal for And<F1, F2>
where
    F1: Filter,
    F2: Filter,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        F1::filter(key, component_map) && F2::filter(key, component_map)
    }
}

impl<F1, F2> FilterSeal for Or<F1, F2>
where
    F1: Filter,
    F2: Filter,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        F1::filter(key, component_map) || F2::filter(key, component_map)
    }
}

impl<C> FilterSeal for &C
where
    C: Component,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        Has::<C>::filter(key, component_map)
    }
}

impl<C> FilterSeal for &mut C
where
    C: Component,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        Has::<C>::filter(key, component_map)
    }
}

impl<C> FilterSeal for Option<&C>
where
    C: Component,
{
    unsafe fn filter(_key: &[u8], _component_map: &HashMap<TypeId, usize>) -> bool {
        true
    }
}

impl<C> FilterSeal for Option<&mut C>
where
    C: Component,
{
    unsafe fn filter(_key: &[u8], _component_map: &HashMap<TypeId, usize>) -> bool {
        true
    }
}

impl FilterSeal for entity::Identifier {
    unsafe fn filter(_key: &[u8], _component_map: &HashMap<TypeId, usize>) -> bool {
        true
    }
}

impl FilterSeal for view::Null {
    unsafe fn filter(_key: &[u8], _component_map: &HashMap<TypeId, usize>) -> bool {
        true
    }
}

impl<'a, V, W> FilterSeal for (V, W)
where
    V: View<'a>,
    W: Views<'a>,
{
    unsafe fn filter(key: &[u8], component_map: &HashMap<TypeId, usize>) -> bool {
        And::<V, W>::filter(key, component_map)
    }
}
