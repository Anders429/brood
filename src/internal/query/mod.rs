pub(crate) mod claim;
pub(crate) mod filter;
#[cfg(feature = "parallel")]
pub(crate) mod par_view;
pub(crate) mod view;
