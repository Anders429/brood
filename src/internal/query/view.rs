use crate::{
    component::Component,
    entity::EntityIdentifier,
    query::{NullResult, NullViews},
};
use core::{any::TypeId, iter, slice};
use hashbrown::HashMap;

pub trait ViewSeal<'a> {
    type Result: Iterator;

    unsafe fn view(
        columns: &[(*mut u8, usize)],
        entity_identifiers: (*mut EntityIdentifier, usize),
        length: usize,
        component_map: &HashMap<TypeId, usize>,
    ) -> Self::Result;
}

impl<'a, C> ViewSeal<'a> for &C where C: Component {
    type Result = slice::Iter<'a, C>;

    unsafe fn view(
        columns: &[(*mut u8, usize)],
        _entity_identifiers: (*mut EntityIdentifier, usize),
        length: usize,
        component_map: &HashMap<TypeId, usize>,
    ) -> Self::Result {
        slice::from_raw_parts::<'a, C>(
            columns
                .get_unchecked(*component_map.get(&TypeId::of::<C>()).unwrap_unchecked())
                .0 as *mut C,
            length,
        )
        .iter()
    }
}

impl<'a, C> ViewSeal<'a> for &mut C
where
    C: Component,
{
    type Result = slice::IterMut<'a, C>;

    unsafe fn view(
        columns: &[(*mut u8, usize)],
        _entity_identifiers: (*mut EntityIdentifier, usize),
        length: usize,
        component_map: &HashMap<TypeId, usize>,
    ) -> Self::Result {
        slice::from_raw_parts_mut::<'a, C>(
            columns
                .get_unchecked(*component_map.get(&TypeId::of::<C>()).unwrap_unchecked())
                .0 as *mut C,
            length,
        )
        .iter_mut()
    }
}

impl<'a> ViewSeal<'a> for EntityIdentifier {
    type Result = iter::Cloned<slice::Iter<'a, Self>>;

    unsafe fn view(
        _columns: &[(*mut u8, usize)],
        entity_identifiers: (*mut EntityIdentifier, usize),
        length: usize,
        _component_map: &HashMap<TypeId, usize>,
    ) -> Self::Result {
        slice::from_raw_parts_mut::<'a, Self>(entity_identifiers.0, length)
            .iter()
            .cloned()
    }
}

pub trait ViewsSeal<'a> {
    type Results: Iterator;

    unsafe fn view(
        columns: &[(*mut u8, usize)],
        entity_identifiers: (*mut EntityIdentifier, usize),
        length: usize,
        component_map: &HashMap<TypeId, usize>,
    ) -> Self::Results;
}

impl<'a> ViewsSeal<'a> for NullViews {
    type Results = iter::Repeat<NullResult>;

    unsafe fn view(
        _columns: &[(*mut u8, usize)],
        _entity_identifiers: (*mut EntityIdentifier, usize),
        _length: usize,
        _component_map: &HashMap<TypeId, usize>,
    ) -> Self::Results {
        iter::repeat(NullResult)
    }
}

impl<'a, V, W> ViewsSeal<'a> for (V, W)
where
    V: ViewSeal<'a>,
    W: ViewsSeal<'a>,
{
    type Results = iter::Zip<V::Result, W::Results>;

    unsafe fn view(
        columns: &[(*mut u8, usize)],
        entity_identifiers: (*mut EntityIdentifier, usize),
        length: usize,
        component_map: &HashMap<TypeId, usize>,
    ) -> Self::Results {
        V::view(columns, entity_identifiers, length, component_map).zip(W::view(
            columns,
            entity_identifiers,
            length,
            component_map,
        ))
    }
}
