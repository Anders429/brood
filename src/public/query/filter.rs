use crate::component::Component;
use core::marker::PhantomData;

pub trait Filter {}

pub struct Has<C>
where
    C: Component,
{
    component: PhantomData<C>,
}

impl<C> Filter for Has<C> where C: Component {}

pub struct Not<F>
where
    F: Filter,
{
    filter: PhantomData<F>,
}

impl<F> Filter for Not<F> where F: Filter {}

pub struct And<F1, F2>
where
    F1: Filter,
    F2: Filter,
{
    filter_1: PhantomData<F1>,
    filter_2: PhantomData<F2>,
}

impl<F1, F2> Filter for And<F1, F2>
where
    F1: Filter,
    F2: Filter,
{
}

pub struct Or<F1, F2>
where
    F1: Filter,
    F2: Filter,
{
    filter_1: PhantomData<F1>,
    filter_2: PhantomData<F2>,
}

impl<F1, F2> Filter for Or<F1, F2>
where
    F1: Filter,
    F2: Filter,
{
}
