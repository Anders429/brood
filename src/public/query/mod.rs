mod filter;
mod view;

pub use filter::*;
pub use view::*;

pub struct Query<V, F> where V: Views, F: Filter {
    views: V,
    filter: F,
}
