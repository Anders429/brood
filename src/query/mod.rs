//! Queries over [`World`]s.
//!
//! Entities within a `World` are difficult to interact with directly due to being made of
//! heterogeneous lists of [`Component`]s. Therefore, queries can be executed to give [`Views`] of
//! `Component`s within the entities stored in a `World`.
//!
//! Queries are made up of `Views`, giving access to `Component`s, and [`Filter`]s which can filter
//! which entities are viewed. Query results are returned as heterogeneous lists, so the
//! [`result!`] macro is provided to unpack the results.
//!
//! # Example
//! The below example queries mutably for the component `Foo`, immutably for the component `Bar`,
//! and filters out entities that do not have the component `Baz`.
//!
//! ``` rust
//! use brood::{
//!     entity,
//!     query::{
//!         filter,
//!         result,
//!         Views,
//!     },
//!     Query,
//!     Registry,
//!     World,
//! };
//!
//! // Define components.
//! struct Foo(u32);
//! struct Bar(bool);
//! struct Baz(f64);
//!
//! type Registry = Registry!(Foo, Bar, Baz);
//!
//! let mut world = World::<Registry>::new();
//! world.insert(entity!(Foo(42), Bar(true), Baz(1.5)));
//!
//! for result!(foo, bar) in world
//!     .query(Query::<Views!(&mut Foo, &Bar), filter::Has<Baz>>::new())
//!     .iter
//! {
//!     // Do something.
//! }
//! ```
//!
//! [`Component`]: crate::component::Component
//! [`Filter`]: crate::query::filter::Filter
//! [`result!`]: crate::query::result!
//! [`Views`]: trait@crate::query::view::Views
//! [`World`]: crate::world::World

pub mod filter;
pub mod result;
pub mod view;

#[doc(inline)]
pub use result::{
    result,
    Result,
};
#[doc(inline)]
pub use view::inner::Views;

use core::{
    fmt,
    marker::PhantomData,
};

/// Defines a query to be run over a world.
///
/// This defines either a regular or parallel query (parallel requires the `rayon` feature to be
/// enabled). It is essentially a marker type, simply providing the types to the calls to
/// [`query()`] to make the API as simple to use as possible.
///
/// # Example
/// ``` rust
/// use brood::{
///     entity,
///     query::{
///         filter,
///         result,
///         Views,
///     },
///     Query,
///     Registry,
///     World,
/// };
///
/// // Define components.
/// struct Foo(u32);
/// struct Bar(bool);
/// struct Baz(f64);
///
/// type Registry = Registry!(Foo, Bar, Baz);
///
/// let mut world = World::<Registry>::new();
/// world.insert(entity!(Foo(42), Bar(true), Baz(1.5)));
///
/// for result!(foo, bar) in world
///     .query(Query::<Views!(&mut Foo, &Bar), filter::Has<Baz>>::new())
///     .iter
/// {
///     // Do something.
/// }
/// ```
///
/// [`query()`]: crate::world::World::query()
pub struct Query<Views, Filters = filter::None, ResourceViews = view::Null> {
    view: PhantomData<Views>,
    filter: PhantomData<Filters>,
    resource_views: PhantomData<ResourceViews>,
}

impl<Views, Filters, ResourceViews> Query<Views, Filters, ResourceViews> {
    /// Creates a new `Query`.
    ///
    /// When creating a query, you must specify the views type `V`, and can optionally specify the
    /// filter type `F`. If no filter is specified, the default filter, [`filter::None`], will be
    /// used.
    #[must_use]
    pub fn new() -> Self {
        Self {
            view: PhantomData,
            filter: PhantomData,
            resource_views: PhantomData,
        }
    }
}

impl<Views, Filters, ResourceViews> Default for Query<Views, Filters, ResourceViews> {
    fn default() -> Self {
        Self::new()
    }
}

impl<Views, Filters, ResourceViews> Clone for Query<Views, Filters, ResourceViews> {
    fn clone(&self) -> Self {
        Self::new()
    }
}

impl<Views, Filters, ResourceViews> PartialEq for Query<Views, Filters, ResourceViews> {
    fn eq(&self, _other: &Self) -> bool {
        true
    }
}

impl<Views, Filters, ResourceViews> Eq for Query<Views, Filters, ResourceViews> {}

impl<Views, Filters, ResourceViews> Copy for Query<Views, Filters, ResourceViews> {}

impl<Views, Filters, ResourceViews> fmt::Debug for Query<Views, Filters, ResourceViews> {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.debug_struct("Query").finish()
    }
}

#[cfg(test)]
mod tests {
    use super::Query;
    use crate::query::Views;
    use alloc::format;

    #[test]
    fn query_default() {
        assert_eq!(Query::<Views!()>::default(), Query::<Views!()>::new());
    }

    #[test]
    fn query_clone() {
        let query = Query::<Views!()>::new();

        assert_eq!(query.clone(), query);
    }

    #[test]
    fn query_debug() {
        assert_eq!(format!("{:?}", Query::<Views!()>::new()), "Query");
    }
}
