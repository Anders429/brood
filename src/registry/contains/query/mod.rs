mod sealed;

use crate::query::view::Views;
use sealed::Sealed;

/// Indicates that a registry is queryable by the filter `F` and the views `V`.
pub trait ContainsQuery<'a, F, FI, V, VI, P, I, Q>: Sealed<'a, F, FI, V, VI, P, I, Q>
where
    V: Views<'a>,
{
}

impl<'a, R, F, V, FI, VI, P, I, Q> ContainsQuery<'a, F, FI, V, VI, P, I, Q> for R
where
    R: Sealed<'a, F, FI, V, VI, P, I, Q>,
    V: Views<'a>,
{
}
