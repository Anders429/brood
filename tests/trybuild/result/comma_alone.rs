use brood::query::{result, view};

fn main() {
    let result!(,) = view::Null;
}
