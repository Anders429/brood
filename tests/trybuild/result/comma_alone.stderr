error: no rules expected the token `,`
 --> tests/trybuild/result/comma_alone.rs:4:17
  |
4 |     let result!(,) = view::Null;
  |                 ^ no rules expected this token in macro call
  |
  = note: while trying to match end of macro
